hypothesis-python/src/hypothesis/searchstrategy/attrs.py

```diff
@@ -0,0 +1,120 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2018 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+from functools import reduce
+from itertools import chain
+from collections import defaultdict
+
+import attr
+
+import hypothesis.strategies as st
+from hypothesis.errors import ResolutionFailed
+from hypothesis.internal.compat import get_type_hints
+from hypothesis.utils.conventions import infer
+
+
+def from_attrs(target, args, kwargs, to_infer):
+    fields = attr.fields(target)
+    kwargs = {k: v for k, v in kwargs.items() if v is not infer}
+    for name in to_infer:
+        kwargs[name] = from_attrs_attribute(getattr(fields, name))
+    # TODO: add a layer here that retries drawing if validation fails, for
+    # improved composition.  Precedent: timezones for datetimes().
+    return st.tuples(st.tuples(*args), st.fixed_dictionaries(kwargs)).map(
+        lambda value: target(*value[0], **value[1])
+    )
+
+
+def from_attrs_attribute(attrib):
+    """Infer a strategy from an attr.Attribute object."""
+    # Various things we know.  Updated below, then inferred to a strategy
+    base = st.nothing()  # updated to none() if None is a possibility
+    default = st.nothing()  # A strategy for the default value, if any
+    in_collections = []  # list of in_ validator collections to sample from
+    # value must be instance of all these types or tuples thereof
+    types = defaultdict(list)  # maps type to list of locations, for error msgs
```

Comment:
I got really confused about what this was doing because of the name. Maybe something like `types_to_locations`?

Reply:
Actually where do we even use this?

Reply:
Er, I actually ended up abandoning the reporting part and this could simply be a set instead.